### PR TITLE
PWDTK refactoring. first 1.x version

### PR DESCRIPTION
**Pull request by galech**

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -9,7 +9,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        python-version: ["3.6", "3.7", "3.8"]
+        python-version: ["3.7", "3.8"]
```
Comment (klausfmh) [inline .github/workflows/build.yml:12]: peut-on ajouter 3.9 et 3.10 ou est-ce-que ca plante qqsparts?

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -9,7 +9,7 @@ jobs:
...
         django-version: ["dj22", "dj31"]
```
Comment (klausfmh) [inline .github/workflows/build.yml:13]: ca marche avec dj40 aussi?

est-ce-que c'est possible de les ajouter a la matrice mais permettre du fail.
Au moins on sait ou on est?

File: pwdtk/auth_backends.py [hunk 3]
```diff
@@ -14,19 +14,39 @@
...
+
+class PwdtkForceRenewException(PwdtkBaseException):
+    """ custom exception """
+
+
 class PwdtkBackend(ModelBackend):
     """
     Authenticates against settings.AUTH_USER_MODEL.
     """
+    def compute_failed_login(self, pwdtk_data):
```
Comment (klausfmh) [inline pwdtk/auth_backends.py:36]: doc strings missing.
mention perhaps et least what data type pwdtk_data is

File: pwdtk/auth_backends.py [hunk 4]
```diff
@@ -14,19 +14,39 @@
...
+        if PwdtkSettings.PWDTK_USER_FAILURE_LIMIT is None:
+            return None
+        if (pwdtk_data.failed_logins+1 >=
+           PwdtkSettings.PWDTK_USER_FAILURE_LIMIT):
+            pwdtk_data.set_locked(pwdtk_data.failed_logins+1)
```
Comment (klausfmh) [inline pwdtk/auth_backends.py:41]: why are line 39 and 40 not on the same line? 
really longer than 80 caracters???

File: pwdtk/auth_backends.py [hunk 5]
```diff
@@ -35,7 +55,7 @@ def authenticate(self, request, username=None, password=None, **kwargs):
             pwdtk_data = PwdData.get_or_create_for_user(user)
             if pwdtk_data.is_locked():
                 raise PwdtkLockedException(pwdtk_data)
-            if user.check_password(password):
+            if (user.check_password(password) and self.user_can_authenticate(user)):
```
Comment (klausfmh) [inline pwdtk/auth_backends.py:58]: why parenthesises if if statement is in one line?

File: pwdtk/auth_backends.py [hunk 6]
```diff
@@ -44,20 +64,17 @@ def authenticate(self, request, username=None, password=None, **kwargs):
                     pwdtk_data.locked = False
                     pwdtk_data.must_renew = must_renew
                     pwdtk_data.save()
+                if must_renew and not kwargs.get("ignore_must_renew"):
+                    raise PwdtkForceRenewException(pwdtk_data)
                 return user
             else:
```
Comment (klausfmh) [inline pwdtk/auth_backends.py:70]: why the else statement.

recommendation if (the if statement contains an unconditianal return, to get rid of the else and unindent the next commands.


```
if condition:
   bla
   return something

do_something_else
```

File: pwdtk/middlewares.py [hunk 7]
```diff
@@ -38,58 +33,26 @@ def __init__(self, get_response=None):
...
-
-        if request.path == reverse(PwdtkSettings.PWDTK_PASSWD_CHANGE_VIEW):
-            return False
-
-        if request.path in PwdtkSettings.PWDTK_PASSWD_CHANGE_ALLOWED_PATHS:
-            return False
-
-        return True
+    def process_request(self, request):
+        if hasattr(PwdtkSettings, "reset_cache"):
```
Comment (klausfmh) [inline pwdtk/middlewares.py:37]: doc string?

what `cache` will be reset?

what is the cache for and why do we have to reset it sometimes

File: pwdtk/middlewares.py [hunk 8]
```diff
@@ -38,58 +33,26 @@ def __init__(self, get_response=None):
...
-        if self.must_renew_password(request):
-            return redirect(reverse(PwdtkSettings.PWDTK_PASSWD_CHANGE_VIEW))
-
-    def process_response(self, request, response):
-
-        if self.must_renew_password(request):
-            return redirect(reverse(PwdtkSettings.PWDTK_PASSWD_CHANGE_VIEW))
+        if isinstance(exception, PwdtkForceRenewException):
+            return HttpResponse(
+                json.dumps({"status": "PWDTK_NEED_RENEW_PASSWORD"}),
```
Comment (klausfmh) [inline pwdtk/middlewares.py:53]: not sure I understand what this means.
PWDTK_NEED_RENEW_PASSWORD


Does this mean we must renew the password to authenticate?

`PWDTK_MUST_RENEW_PASSWORD`

or

`PWDTK_NEEDS_RENEW_PASSWORD` (NEEDSA with an S)

File: tox.ini [hunk 9]
```diff
@@ -1,5 +1,5 @@
 [tox]
-envlist=py{36,37,38,39}-django{22,31}
+envlist=py{37,38,39}-django{22,31}
```
Comment (klausfmh) [inline tox.ini:2]: make sure versions are identical for gitlab / github

File: setup.py [hunk 10]
```diff
@@ -50,6 +50,6 @@
     },
     # TODO in the near future
     # python_requires=">=3.6, <4",
-    python_requires=">=3.6, <4",
+    python_requires=">=3.7, <4",
```
Comment (klausfmh) [inline setup.py:53]: If you have time you might start adding a pyproject.toml file 
as it is recommended now (you can check with Qeuntin or me  or look at other repos)

I think pypeman, mh_lib and others sue alerady pyproject.toml

File: pwdtk/settings.py [hunk 11]
```diff
@@ -39,6 +39,8 @@ def add_middlewares(
 # PWDTK Passsword lockout after logins with bad password
 # ----------------------------------------------------------
 
+PWDTK_RAISE_MUST_RENEW_CLASSES = ["rest_framework.request.Request", ]
```
Comment (klausfmh) [inline pwdtk/settings.py:42]: comment to explain what `PWDTK_RAISE_MUST_RENEW_CLASSES ` means.
It is not obvious for me.

I guess it is something, that only rest requests will raise renew exceptions but not normal requests.

but I only guess this as I once overheard you talking about something like this

File: .github/workflows/build.yml (hunk 1 above)
Comment (galech) [reply .github/workflows/build.yml:12]: je ne sais pas, je vais ajouter et on regarde ce que ca donne

File: pwdtk/middlewares.py (hunk 7 above)
Comment (galech) [reply pwdtk/middlewares.py:37]: shit, resolved dont save comments

Comment (galech) [reply pwdtk/middlewares.py:37]: Te goal is to allow live pwdtk settings. Sometimes setting could come from other sources like DB entries.
We keep a cache only for one request instance, thsi allow change parameters or activate/deactivate pwdtk whit not restart required

File: pwdtk/middlewares.py (hunk 8 above)
Comment (galech) [reply pwdtk/middlewares.py:53]: it is more like NEEDS, wrong typo but maybe we can keep this

File: pwdtk/settings.py (hunk 11 above)
Comment (galech) [reply pwdtk/settings.py:42]: It is obsolete, i will remove

File: pwdtk/auth_backends.py (hunk 6 above)
Comment (galech) [reply pwdtk/auth_backends.py:70]: will do

File: pwdtk/middlewares.py (hunk 8 above)
Comment (asixela) [reply pwdtk/middlewares.py:53]: PWDTK_NEED_RENEW_PASSWORD status is already used

File: setup.py (hunk 10 above)
Comment (galech) [reply setup.py:53]: no time ;)

File: pwdtk/auth_backends.py (hunk 3 above)
Comment (galech) [reply pwdtk/auth_backends.py:36]: done

**Pull request merged**